File: Server_Root/Server/RoomServer/PlayerInput/PlayerInputs/Position.cpp
#include "stdafx.h"
#include "Position.h"
#include "../../../FBSFiles/PlayerInput/Position_generated.h"

namespace PlayerInput
{
	Position::Position()
	{
	}

	Position::~Position()
	{
	}

	FBS::PlayerInputType Position::GetType()
	{
		return FBS::PlayerInputType::PlayerInputType_Position;
	}

	IPlayerInput* Position::Clone()
	{
		Position* pClone = new Position();

		pClone->m_nEntityID = m_nEntityID;
		pClone->m_vec3Position = m_vec3Position;

		return pClone;
	}

	const char* Position::Serialize(int* pLength)
	{
		FBS::Data::Vector3 position(m_vec3Position.x(), m_vec3Position.y(), m_vec3Position.z());

		FBS::PlayerInput::PositionBuilder data_builder(m_Builder);
		data_builder.add_EntityID(m_nEntityID);
		data_builder.add_Pos(&position);
		auto data = data_builder.Finish();

		m_Builder.Finish(data);

		*pLength = m_Builder.GetSize();

		return (char*)m_Builder.GetBufferPointer();
	}

	bool Position::Deserialize(const char* pChar)
	{
		auto data = flatbuffers::GetRoot<FBS::PlayerInput::Position>((const void*)pChar);

		m_nEntityID = data->EntityID();
		m_vec3Position.setValue(data->Pos()->x(), data->Pos()->y(), data->Pos()->z());

		return true;
	}
}

File: Client_Root/Client/Assets/Scripts/PlayerInput/PlayerInputs/Position.cs
using UnityEngine;
using FlatBuffers;

namespace PlayerInput
{
    public class Position : IPlayerInput
    {
        public int      m_nEntityID;
        public Vector3  m_vec3Position;

        public override FBS.PlayerInputType GetPlayerInputType()
        {
            return FBS.PlayerInputType.Position;
        }

        public override byte[] Serialize()
        {
            var pos = FBS.Data.Vector3.CreateVector3(m_Builder, m_vec3Position.x, m_vec3Position.y, m_vec3Position.z);

            FBS.PlayerInput.Position.StartPosition(m_Builder);
            FBS.PlayerInput.Position.AddEntityID(m_Builder, m_nEntityID);
            FBS.PlayerInput.Position.AddPos(m_Builder, pos);
            var data = FBS.PlayerInput.Position.EndPosition(m_Builder);

            m_Builder.Finish(data.Value);

            return m_Builder.SizedByteArray();
        }

        public override bool Deserialize(byte[] bytes)
        {
            var buf = new ByteBuffer(bytes);

            var data = FBS.PlayerInput.Position.GetRootAsPosition(buf);

            m_nEntityID = data.EntityID;
            m_vec3Position = new Vector3(data.Pos.X, data.Pos.Y, data.Pos.Z);

            return true;
        }

        public override string ToString()
        {
            return string.Format("EntityID : {0}, Position : {1}", m_nEntityID, m_vec3Position.ToString());
        }
    }
}